### PR TITLE
feat(grpc): drive vLLM MooncakeConnector PD with router-minted kv_transfer_params

### PR DESCRIPTION
**Pull request by CatherineSue**
### Problem

SMG's vLLM PD Mooncake path injects only `remote_host`/`remote_port` into the decode request. The modern in-tree `MooncakeConnector` (vLLM ≥ 0.20-era) ignores those keys entirely: it is push-based, requires `do_remote_decode` + a `transfer_id` on the prefill request (else it pins nothing and warns "Missing transfer_id in kv_transfer_params from router!"), and requires `remote_engine_id`/`remote_bootstrap_addr`/`transfer_id` on the decode request. Unlike NIXL, `request_finished()` returns no params — there is nothing to harvest from the prefill response — so the relay flow shipped in #1622 cannot drive it. Net effect: against upstream vLLM, Mooncake PD is a silent no-op (decode recomputes the prompt), the same bug class #1622 fixed for NIXL. There is also no CI coverage for vLLM-Mooncake PD.

### Solution

Implement the contract the ecosystem converged on — verified against four independent implementations (vLLM's own example proxy at the v0.22.1 tag, vllm-project/router, NVIDIA Dynamo PR #9414, llm-d's routing sidecar), all of which mint and synthesize the params outside the engine:

1. **Discovery**: the servicer reports `kv_transfer_config.engine_id` via a new `GetServerInfoResponse.kv_engine_id` field, stored in worker metadata at registration alongside `kv_connector`/`kv_role` (smg-grpc-proto 0.4.10, smg-grpc-servicer 0.5.5).
2. **Router**: in Mooncake mode the router mints a per-attempt `transfer_id`, tags the prefill request `{do_remote_decode: true, transfer_id}`, and synthesizes the decode params `{do_remote_prefill: true, transfer_id, remote_engine_id, remote_bootstrap_addr: "http://bootstrap_host:bootstrap_port"}`. The NIXL relay path and the n=1 guard are unchanged; legacy typed host/port injection remains the fallback whenever `engine_id` is unknown.
3. **Safety gates** (from adversarial review): the servicer reports no engine id when vLLM data parallelism is active (the engine cores suffix their connector ids per rank, so a minted id would make decode hang instead of degrade) — DP-aware rank pinning is a documented follow-up; empty-string `kv_engine_id` overrides force the legacy fallback instead of synthesizing broken params.
4. **CI**: new `e2e-2gpu-pd (vllm-mooncake)` matrix leg (`E2E_VLLM_KV_BACKEND` parametrization, tcp transport to stay off the flaky shared-node RDMA path, producer bootstrap server on the advertised port). `test_pd_mooncake.py` asserts the router injected minted params and the workers logged a **positive** successful-transfer count (the metrics line also prints at zero when every transfer fails). A fatal import canary catches broken `mooncake-transfer-engine` installs at setup instead of mid-e2e, and failure artifacts are disambiguated per KV backend.

Operational notes (documented): pin a stable `engine_id` in the prefill worker's `--kv-transfer-config` — SMG discovers the id once at registration, and vLLM generates a fresh uuid per process, so an unpinned restarted container invalidates the registered id. Metadata refresh on worker recovery (useful beyond Mooncake) and DP-aware routing are planned as separate follow-up PRs.

## Changes

- `crates/grpc_client/proto/vllm_engine.proto`: `GetServerInfoResponse.kv_engine_id = 8`; smg-grpc-proto → 0.4.10.
- `grpc_servicer`: report `kv_engine_id` (empty when DP is active); smg-grpc-servicer → 0.5.5.
- `crates/protocols` + `model_gateway` worker/discovery: `WorkerSpec.kv_engine_id` plumbed from discovery labels (empty-string filtered).
- `model_gateway` router: Mooncake mint/tag/synthesize with legacy fallback; unit tests for mode mapping, param shapes, and the empty-id fallback.
- `e2e_test`: KV-backend parametrization in worker launch, shared `infra/pd_logs.py` helpers, `test_pd_mooncake.py`, backend skip-gates on the NIXL test.
- `.github/workflows` + `scripts/ci_install_vllm.sh`: vllm-mooncake PD matrix leg, `mooncake-transfer-engine` install + import canary, per-backend artifact names.
- `docs`: Mooncake flow description, `engine_id` pinning guidance, DP limitation.

## Test Plan

- `cargo clippy --all-targets` clean; `cargo test -p smg --lib`: 1015 passed (new tests: Mooncake mode mapping incl. empty/missing engine_id fallbacks, prefill/decode param JSON shapes).
- `pytest grpc_servicer/tests`: 20 passed.
- New `e2e-2gpu-pd (vllm-mooncake)` job on this PR: asserts minted-param injection in router logs and `Num successful transfers=[1-9]` in worker logs — the assertion fails if decode silently recomputes.
- Existing `e2e-2gpu-pd (vllm)` NIXL leg guards the unchanged relay path; `(sglang)` leg unaffected.

<details>
<summary>Checklist</summary>

- [x] `cargo +nightly fmt` passes
- [x] `cargo clippy --all-targets --all-features -- -D warnings` passes
- [x] (Optional) Documentation updated
- [ ] (Optional) Please join us on Slack [#sig-smg](https://slack.lightseek.org) to discuss, review, and merge PRs

</details>

## Summary by CodeRabbit

* **New Features**
  * Mooncake KV-transfer support for vLLM PD with runtime backend selection; router can mint/inject transfer params and expose worker engine IDs.

* **Documentation**
  * Expanded PD disaggregation guides covering Mooncake flow, engine_id pinning, DP limitations, and fallback behavior.

* **Tests**
  * New/updated e2e tests for Mooncake and Nixl paths; shared log utilities and log-based assertions added.

* **Chores**
  * CI/workflow and package metadata updates; CI can optionally install Mooncake runtime.

### CODE REVIEW

File: scripts/ci_install_vllm.sh [hunk 1]
```diff
@@ -44,6 +44,15 @@ rm -rf "${SITE_PACKAGES}/nixl_ep"
 python3 -c "import torch, nixl"
 echo "nixl import canary OK"
 
+# Mooncake transfer engine for the vLLM MooncakeConnector PD leg
+echo "Installing mooncake-transfer-engine..."
+uv pip install "mooncake-transfer-engine>=0.3.8"
+
+# Import canary: fail here (not mid-e2e) if the mooncake install is broken —
+# vLLM swallows this ImportError at module load (torch first for CUDA libs)
+python3 -c "import torch; from mooncake.engine import TransferEngine"
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply scripts/ci_install_vllm.sh]: Done in 1d3fa0f9 — the install and canary are now gated on E2E_VLLM_KV_BACKEND=mooncake, so only the mooncake PD leg depends on the wheel; the other vLLM jobs no longer share that failure mode.

File: model_gateway/src/routers/grpc/common/stages/request_execution.rs [hunk 2]
```diff
@@ -43,17 +49,43 @@ fn kv_connector_mode(
...
+/// Decode-leg params for Mooncake, synthesized from prefill worker metadata
+/// (the engine returns nothing to relay; the connector is push-based).
+fn mooncake_decode_params(transfer_id: &str, engine_id: &str, host: &str, port: u32) -> String {
+    serde_json::json!({
+        "do_remote_decode": false,
+        "do_remote_prefill": true,
+        "transfer_id": transfer_id,
+        "remote_engine_id": engine_id,
+        "remote_bootstrap_addr": format!("http://{host}:{port}"),
+    })
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply model_gateway/src/routers/grpc/common/stages/request_execution.rs:85]: Defer, we can do this later, not critical right now.

File: e2e_test/router/test_pd_mooncake.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        )
+        assert response.choices[0].message.content
+
+        worker_dir = worker_log_dir(_LOG_DIR)
+        worker_logs = wait_for_pattern(worker_dir, "worker-*.log", TRANSFER_SUCCESS_RE)
+        if not worker_logs:
+            pytest.skip(
+                "No worker log files captured (SHOW_WORKER_LOGS=1?); "
+                "cannot assert on Mooncake transfer"
+            )
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply e2e_test/router/test_pd_mooncake.py]: Done in d1274f05 — empty worker logs now fail the test when CI=true and skip only locally (where SHOW_WORKER_LOGS=1 legitimately routes logs to the console). Applied via a shared assert_worker_logs_captured helper in infra/pd_logs.py and used by both test_pd_mooncake and test_pd_nixl, so the NIXL leg gets the same hardening.

**Comment by slin1237** [comment]
Investigated the failing `Setup vLLM backend` step and pushed `13167d77` with the fix.

**Root cause:** the import canary died on `libibverbs.so.1: cannot open shared object file` — mooncake's native extension links libibverbs/libnuma at **load** time even with `mooncake_protocol=tcp`, and the vLLM pods don't have the RDMA userspace libraries (the sglang install script apt-installs them right before mooncake; the new vLLM block didn't). The fix adds `libnuma1 libibverbs1 ibverbs-providers`.

**Two related cleanups in the same commit:**
- Switched to the `mooncake-transfer-engine-cuda13` wheel variant, which matches vLLM's cu130 torch stack — this deletes the `nvidia-cuda-runtime-cu12` / `libcudart.so.12` `LD_LIBRARY_PATH` shim entirely (torch's bundled CUDA 13 runtime satisfies the link once torch is imported first, which the canary already does), and stops persisting a cu12 library path into `GITHUB_ENV` for a cu13 job. Same variant sglang upstream uses on cu13 and that #1662 adopted.
- Pinned `==0.3.11.post1` (the release the floating `>=0.3.8` resolved to anyway) — an unpinned transitive resolve is exactly what broke the sglang lane this week (`kernels` 0.15).

The `mooncake_protocol: tcp` choice in the e2e worker config is a good call — it also sidesteps the RDMA memlock issue currently affecting the rotated CI GPU nodes (see #1574-adjacent discussion / the sglang PD lane).
